src/main/java/com/amazon/opendistroforelasticsearch/jdbc/ResultSetImpl.java

```diff
@@ -112,15 +123,56 @@ public boolean next() throws SQLException {
         log.debug(() -> logEntry("next()"));
         checkOpen();
         boolean next = cursor.next();
+
+        if (!next && this.cursorId != null) {
```

Comment:
This logic should be encapsulated in Cursor.

Reply:
Why should that be the case? Also doing so will require passing `Statement` reference to `Cursor` in order  to get `Connection` object in . It's doable, but I feel it does not logically belongs there.

I believe the `Cursor` object just hold the rows and schema. The logic of fetching next set of rows should be handled by `ResultSet` and not delegated to `Cursor`. Just my thought. I am open to suggestion.

Reply:
You are right, Cursor should hold the rows, Then every time call Cursor.next(), should return the next row on the dataset. This is why I suggested to encapsulated cursorId related logic in Cursor. If the cursor.next() == false, IMO it is the Cursor's responsibility to fetch additional rows if possible.
In addition, the Cursor should be final. The cursor in L125 and L131 is different Cursor. So the cursor.next been called two times.
Ideally, the Cursor should be injected to the ResultSetImpl instead of construct inside it. The major benefit of using dependency injection is we can easily mock the Cursor and then test ResultSetImpl.
By the way, if the change require a lot of effort and impact the release, I suggest to add some comments.

Reply:
Add comments for now.